### PR TITLE
issue_34282 - Missing "</form>" for the form_with and form_tag helpers if they don't have a html block

### PR DESCRIPTION
**Pull request by OvidiuToma**
If `form_with` and `form_tag` were called without a html block, those helpers displayed a html form without ending in `</form>`. 
This issue caused a bug where a form is not displayed if it succeeds a form_with/form_tag that doesn't have a html block.

### Summary
Fixes issue: https://github.com/rails/rails/issues/34282

**Code:**
screen shot 2018-10-22 at 16 55 26

**Before(second form is not displayed):**
screen shot 2018-10-22 at 17 54 10

**After:**
screen shot 2018-10-22 at 16 55 16

### Other Information

- Executed the tests for the actionview and they seem fine.
- Added an item to the actionview/CHANGELOG.md file.

### CODE REVIEW

**Review by gmcgibbon — COMMENTED** [review]
You'll need to add a test for this to make sure this bug doesn't regress. Still, I'm curious if its better to just raise an error instead? What's a use-case where you would want a form with no body?

**Comment by OvidiuToma** [comment]
@gmcgibbon Raising an error won't do in my opinion. Both of the usages are syntactically valid, as it can be seen in form_tag [documentation](https://apidock.com/rails/ActionView/Helpers/FormTagHelper/form_tag)(check examples). 

```
form_tag('/posts')
# => <form action="/posts" method="post">

form_tag('/posts/1', method: :put)
# => <form action="/posts/1" method="post"> ... <input name="_method" type="hidden" value="put" /> ...

form_tag('/upload', multipart: true)
# => <form action="/upload" method="post" enctype="multipart/form-data">

<%= form_tag('/posts') do -%>
  <div><%= submit_tag 'Save' %></div>
<% end -%>
# => <form action="/posts" method="post"><div><input type="submit" name="commit" value="Save" /></div></form>

<%= form_tag('/posts', remote: true) %>
# => <form action="/posts" method="post" data-remote="true">

form_tag('http://far.away.com/form', authenticity_token: false)
# form without authenticity token

form_tag('http://far.away.com/form', authenticity_token: "cf50faa3fe97702ca1ae")
# form with custom authenticity token
```

**Comment by lsylvester** [comment]
I think that this intentional. In the docs at https://github.com/rails/rails/blob/b13a5cb83ea00d6a3d71320fd276ca21049c2544/actionview/lib/action_view/helpers/form_tag_helper.rb#L27 it describes it as "Starts a form tag", and it provides an example at https://github.com/rails/rails/blob/b13a5cb83ea00d6a3d71320fd276ca21049c2544/actionview/lib/action_view/helpers/form_tag_helper.rb#L48-L49 where this only an opening tag.

I think that this was to allow using the form helper then the form starts and ends in different templates. 

I think that if this is going to change the docs would need to be updated and it would need to be deprecated. 

Is there a use case for a empty form? 

I think that requiring a block might be the way to go. If we want to be able to start a form without an end (to close elsewhere in the code) it might be better as a `start_form_tag` helper or similar - so it is clearer it is not the whole form - and it needs to be finished elsewhere.

**Comment by rafaelfranca** [comment]
This is definitely by design, and this is not something we want to remove support of in `form_tag`.

`form_tag` is is a low level input helper. The low level input helpers has as idea to maps the HTML elements as closer as possible without much magic. 

`form_with` is a different story. I'm fine with changing it and its documentation to always include the `</form>`.

**Comment by sham-sheer** [comment]
sda

**Comment by kaspth** [comment]
@rafaelfranca form_with can also be used like form_tag. That’s part of it’s design. So I wouldn’t change the docs here.

**Comment by rafaelfranca** [comment]
Yeah, I get that but `form_with` doesn't need to be handled as low level API as `form_tag`. It is for sure smarter than `form_tag` so in that case it makes sense to always generate the `</form>` tag even if the block is not passed.

**Comment by kaspth** [comment]
@rafaelfranca yeah, that makes sense to me 👍

File: actionview/test/template/form_helper/form_with_test.rb [hunk 1]
```diff
@@ -2319,6 +2319,11 @@ def test_form_with_only_instantiates_builder_once
     assert_equal 1, initialization_count, "form builder instantiated more than once"
   end
 
+  def test_if_form_with_displays_form_that_is_closed
```
[the remark this answers is not in the document]
Comment (OvidiuToma) [reply actionview/test/template/form_helper/form_with_test.rb]: Thanks for the suggestion! Will update.

**Comment by matthewd** [comment]
Do we want to make it output an empty form, or do we want it to raise / treat the block as required?

**Comment by gmcgibbon** [comment]
I just found https://github.com/rails/rails/blob/3f8b1353c64825fb665f3ab0c06d3616dc56b6fe/actionview/lib/action_view/helpers/form_helper.rb#L626
We will also have have to update this if the behaviour changes.

**Comment by OvidiuToma** [comment]
First of all, thank you all for commenting.

As it can be seen by your comments, according to the docs and code comments, it looks like the expected behaviour is the current one. As in, if a code block is not passed to those helpers, it will not generate a closing tag. 

However, this can lead to unexpected behaviour in your html. I've made some more tests and the second form can display it's html fields tags without being included in a form tag because of this "expected" behaviour. **So it's not just an issue that the second form is not displayed.** 

screen shot 2018-10-26 at 12 38 20
screen shot 2018-10-26 at 12 36 51

I can agree that this behaves this way so it gives the developer the flexibility to end the form wherever he wants but it just seems a little too "hack-ish" and not the rails way:
screen shot 2018-10-26 at 12 56 33


That being said, I would agree that we could raise an error, but isn't that too aggressive?

**Comment by koenhandekyn** [comment]
Actually just to "open" this again, in one of the apps a friend is working on, "we" hit this same 'surprise' again. (https://github.com/rails/rails/issues/34282) 

This was a good summary by "[OvidiuToma](https://github.com/OvidiuToma)" in my opinion: "seems a little too "hack-ish" and not the rails way" (to not close the form when no block is passed).

**Comment by seanpdoyle** [comment]
https://github.com/rails/rails/pull/44275 proposes a potential resolution to this issue.

**Pull request closed without merging**